# 4

In [2]:
import numpy as np

P = np.array([
    [0, 1, 0, 0],
    [1/3, 0, 2/3, 0],
    [0, 0, 0, 1],
    [1/2, 0, 1/2, 0]
])

def evolve(p, P, n):
    for _ in range(n):
        p = p @ P
    return p


In [4]:
p0 = np.array([1/4, 1/4, 1/4, 1/4])

for n in range(0, 20):
    print(f"p_{n} =", evolve(p0, P, n))


p_0 = [0.25 0.25 0.25 0.25]
p_1 = [0.20833333 0.25       0.29166667 0.25      ]
p_2 = [0.20833333 0.20833333 0.29166667 0.29166667]
p_3 = [0.21527778 0.20833333 0.28472222 0.29166667]
p_4 = [0.21527778 0.21527778 0.28472222 0.28472222]
p_5 = [0.21412037 0.21527778 0.28587963 0.28472222]
p_6 = [0.21412037 0.21412037 0.28587963 0.28587963]
p_7 = [0.21431327 0.21412037 0.28568673 0.28587963]
p_8 = [0.21431327 0.21431327 0.28568673 0.28568673]
p_9 = [0.21428112 0.21431327 0.28571888 0.28568673]
p_10 = [0.21428112 0.21428112 0.28571888 0.28571888]
p_11 = [0.21428648 0.21428112 0.28571352 0.28571888]
p_12 = [0.21428648 0.21428648 0.28571352 0.28571352]
p_13 = [0.21428559 0.21428648 0.28571441 0.28571352]
p_14 = [0.21428559 0.21428559 0.28571441 0.28571441]
p_15 = [0.21428574 0.21428559 0.28571426 0.28571441]
p_16 = [0.21428574 0.21428574 0.28571426 0.28571426]
p_17 = [0.21428571 0.21428574 0.28571429 0.28571426]
p_18 = [0.21428571 0.21428571 0.28571429 0.28571429]
p_19 = [0.21428571 0.214285

Yes, when initialized at $p_0 = [1/4,1/4,1/4,1/4]$, the chain appears to convege near $p_n = [.21428571, .21428571, 0.28571429, 0.28571429]$.

In [8]:
p0 = np.array([1, 0, 0, 0])

for n in range(0, 20):
    print(f"p_{n} =", evolve(p0, P, n))


p_0 = [1 0 0 0]
p_1 = [0. 1. 0. 0.]
p_2 = [0.33333333 0.         0.66666667 0.        ]
p_3 = [0.         0.33333333 0.         0.66666667]
p_4 = [0.44444444 0.         0.55555556 0.        ]
p_5 = [0.         0.44444444 0.         0.55555556]
p_6 = [0.42592593 0.         0.57407407 0.        ]
p_7 = [0.         0.42592593 0.         0.57407407]
p_8 = [0.42901235 0.         0.57098765 0.        ]
p_9 = [0.         0.42901235 0.         0.57098765]
p_10 = [0.42849794 0.         0.57150206 0.        ]
p_11 = [0.         0.42849794 0.         0.57150206]
p_12 = [0.42858368 0.         0.57141632 0.        ]
p_13 = [0.         0.42858368 0.         0.57141632]
p_14 = [0.42856939 0.         0.57143061 0.        ]
p_15 = [0.         0.42856939 0.         0.57143061]
p_16 = [0.42857177 0.         0.57142823 0.        ]
p_17 = [0.         0.42857177 0.         0.57142823]
p_18 = [0.42857137 0.         0.57142863 0.        ]
p_19 = [0.         0.42857137 0.         0.57142863]


There seems to be an oscillation between two modes (even/odd steps), especially at later steps, where all the mass is on either the even indices or the odd indices. The chain seems to be converging to an oscillatory sequence. To confirm, we check larger time steps of the chain:

In [9]:
for n in range(400, 420):
    print(f"p_{n} =", evolve(p0, P, n))

p_400 = [0.42857143 0.         0.57142857 0.        ]
p_401 = [0.         0.42857143 0.         0.57142857]
p_402 = [0.42857143 0.         0.57142857 0.        ]
p_403 = [0.         0.42857143 0.         0.57142857]
p_404 = [0.42857143 0.         0.57142857 0.        ]
p_405 = [0.         0.42857143 0.         0.57142857]
p_406 = [0.42857143 0.         0.57142857 0.        ]
p_407 = [0.         0.42857143 0.         0.57142857]
p_408 = [0.42857143 0.         0.57142857 0.        ]
p_409 = [0.         0.42857143 0.         0.57142857]
p_410 = [0.42857143 0.         0.57142857 0.        ]
p_411 = [0.         0.42857143 0.         0.57142857]
p_412 = [0.42857143 0.         0.57142857 0.        ]
p_413 = [0.         0.42857143 0.         0.57142857]
p_414 = [0.42857143 0.         0.57142857 0.        ]
p_415 = [0.         0.42857143 0.         0.57142857]
p_416 = [0.42857143 0.         0.57142857 0.        ]
p_417 = [0.         0.42857143 0.         0.57142857]
p_418 = [0.42857143 0.      

Clearly, the distribution converges to an oscllatory distribution, where it is either on an even mode/step (then, with distribution of [0.42857143, 0. , 0.57142857 0.,]) or an odd mode (with distribution of [0., 0.42857143, 0. ,0.57142857]). Note that the probability distribution is discrete (e.g., only mass on the 1st/3rd states or the 2nd/4th states), and the transition matrix shifts the mass either to the right or left.

We do not get the same limit as with an  initial distribution  of $p_0 = [1/4,1/4,1/4,1/4]$, which converges to a stationary point. Instead we get a distribution with a period of 2.